### PR TITLE
New shot handling backbone

### PR DESCRIPTION
**Pull request by Liruxo**
First of all: I am sorry, this is quite a big pull request because i am unable to split this into multiple PRs due to the nature of this project
Second this also fixes multiple issues, because they are all related to the old shooting code

The Content of this PR:
- Penetration is now working
- Everything that shoots now uses the new unified code, therefore multiple Entitys with guns can now perform raytraces based on the properties of the gun
-The raytrace is now always performed on the server side

Problems:
-Mechs currently cant use grenade launching guns, because this would require a rewrite of the grenade related classes
-There may be some "unsmooth" playing experience due to packet delay over the internet, which is a problem since now the client needs to wait for the server to make hit related effects and sounds
-GunTrails are not visible when standing to close to the origin of the shot

Questions:
-To counter the second Problem i have implemented a Queue (found in the PlayerData class) to delay shoots until they should be shot. This however results in constantly resetting the magazine on the client side, because for the player all this shots were actually fired with the correct timings. The current solution is to skip ticks on the server side and just trust that the client is using the correct timings.
The question is how this should be handled
-GunTrails are currently visible to all players, which wasnt possible before because no Packet to realize this existed. Should GunTrails be visible to all players?

### CODE REVIEW

**Comment by ChrisLane** [comment]
Hi, thanks for contributing such a large change.

I've not conducted a full review or test of the code yet since I had some issues with the problems you stated.
I don't want to merge in code that adds new bugs, therefore the mecha grenade launcher shooting needs to be fixed.

"GunTrails are not visible when standing to close to the origin of the shot" - Does this mean that players firing the guns can't see the trails? That seems like it's a fairly large bug in the feature.

**Comment by Liruxo** [comment]
I have conducted a few test: The GunTrail issue is not related to my changes in any way, in fact it exists for a very long time. For the grenade launcher: this bug also existed before my changes.
I have reported both bugs to the Issue section (#1022 & #1023) , since both work in the legacy versions.

File: .gitignore
Comment (ChrisLane) [inline .gitignore]: Unrelated change

File: src/main/java/com/flansmod/apocalypse/common/entity/EntityFlansModShooter.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/apocalypse/common/entity/EntityFlansModShooter.java]: A crash is definitely worse than the guns not working at all.

File: src/main/java/com/flansmod/common/PlayerData.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/PlayerData.java]: Remove debug prints, please.

File: src/main/java/com/flansmod/common/PlayerData.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/PlayerData.java]: Nope, if you check the definitions of the methods, they'll return an ItemStack.Empty instead of null.

File: src/main/java/com/flansmod/common/PlayerData.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/PlayerData.java]:
```suggestion
			FlansMod.log.warn("Dropping client shoot requests because the client is out of sync with the server by more than half a second");
```

File: src/main/java/com/flansmod/common/driveables/EntitySeat.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/driveables/EntitySeat.java]: Remove debug prints

File: src/main/java/com/flansmod/common/driveables/EntitySeat.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/driveables/EntitySeat.java]: Remove unused code.

File: src/main/java/com/flansmod/common/guns/FireableGun.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/FireableGun.java:8]: Is there the need for this new class when a lot of the data is already stored in `GunType`?

File: src/main/java/com/flansmod/common/guns/ItemGun.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/ItemGun.java:433]: I'd rather this be a positive method i.e. `gunCanBeHandeled`

File: src/main/java/com/flansmod/common/guns/ItemGun.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/ItemGun.java:554]: I don't know what this TODO means

File: src/main/java/com/flansmod/common/guns/ItemGun.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/ItemGun.java:608]: Method names should be camel case `playShotSound`

File: src/main/java/com/flansmod/common/guns/FiredShot.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/FiredShot.java:14]: Seems this class could have just become `ShotData`. Can a shot not be fired? :wink:

File: src/main/java/com/flansmod/common/guns/ShotHandler.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/ShotHandler.java:149]: Gun trails should be visible to all players so it seems that this todo can be removed.

File: src/main/java/com/flansmod/common/guns/ShotHandler.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/ShotHandler.java:162]: Camel case, even for static methods

File: src/main/java/com/flansmod/common/guns/raytracing/FlansModRaytracer.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/raytracing/FlansModRaytracer.java:148]: Added commented code?

File: src/main/java/com/flansmod/common/guns/raytracing/FlansModRaytracer.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/guns/raytracing/FlansModRaytracer.java:259]: Can use the primitive `float` type here.

File: src/main/java/com/flansmod/common/network/PacketBlockHitEffect.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/network/PacketBlockHitEffect.java:47]: Use primitive types where possible please, no need for `Float` and `Integer`

File: src/main/java/com/flansmod/common/network/PacketBlockHitEffect.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/network/PacketBlockHitEffect.java:36]:
```suggestion
	private EnumFacing facing;
```

File: src/main/java/com/flansmod/common/network/PacketBulletTrail.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/network/PacketBulletTrail.java]: Remove todo, use `FlansMod.log.warn`

File: src/main/java/com/flansmod/common/network/PacketGunAnimation.java
Comment (ChrisLane) [inline src/main/java/com/flansmod/common/network/PacketGunAnimation.java:183]: Replace code like this with a switch statement

**Review by ChrisLane — CHANGES_REQUESTED** [review]
I've skimmed through the code and there are some cleanup changes to be made. Please act on the comments that I've made, some apply to many places in the changes but the comments only appear once.
I can see that you've added a significant amount of TODOs. If you could try to resolve as many of the TODOs as possible, especially the ones related to your changes, that would be great.

I'll be in-game testing when I get the time.

File: src/main/java/com/flansmod/apocalypse/common/entity/EntityFlansModShooter.java
Comment (Liruxo) [reply src/main/java/com/flansmod/apocalypse/common/entity/EntityFlansModShooter.java]: The crash is complete speculation. I am currently unable to enter the apocalypse world and/or spawn in an Entity that is using this code. Therefore i cant verify nor deny that a crash would be caused by this code. However i added code to prevent any possible crash

**Comment by ChrisLane** [comment]
I see some of my recent commits have come through as co-authored by you. Please merge next time to avoid confusing the commit tree.

**Comment by Liruxo** [comment]
Thats weird, because i did merge. Well i try to avoid it but i am unsure what caused GitHub to add me as Co-Author

File: src/main/java/com/flansmod/common/guns/FireableGun.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/guns/FireableGun.java:8]: GunType is a specific data storage for guns which exists as items. This makes it impossible to use GunType as general data storage because for example AAGunType is completely different class which cant be converted into GunType.
FireableGun unifys all the important data about the gun, making the code using FireableGun very simple because the acquiring of the data is done by the code which calls the fireGun method

File: src/main/java/com/flansmod/common/guns/ShotHandler.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/guns/ShotHandler.java:162]: Just copied it over from the old code :smile:

File: src/main/java/com/flansmod/common/guns/FiredShot.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/guns/FiredShot.java:14]: They share a lot of similarity's, FiredShot is just a complete rebuild with its new function in mind, which is to convert the many informations in a standardized format

**Comment by Liruxo** [comment]
I now have implemented most of your suggestions and marked the conversation as resolved. The remaining conservations all have an comment and are waiting for your response/approval/critic
I am currently still working on some EntityBullet related speed issues, but i am currently also tight on time so ill write another message when i am finished (It will probably be a longer period)

**Comment by arolypoly** [comment]
Merged and built recently.
With mods:
https://gist.github.com/rolylolypoly/bc024262d497aebba12eb90869ce08b7
Without mods:
https://gist.github.com/rolylolypoly/8a7765dde40813232b6ccbdf339a8988

**Comment by arolypoly** [comment]
Running it with gradle runClient on Linux = no issues.
Running the build/output on Windows 10 causes it to hang in strange ways (I cannot get the code to emit a log message before it completely hangs).

**Comment by Liruxo** [comment]
I have now fixed numerous issues related to unloading/loading EntityBullet's, their trajectory, client/server desynchronization, etc. I am now finished as everything i tested related to my changes works as intended.
I've tested:

-   Driveables:
    1. Passengerguns
    2. Driverguns: Bomb/Rockets & Tank Shells & Miniguns etc.
-  Inventory Guns:
   1.  Usual guns, e.g. Ak47
   2. Minigun
   3. Grenade Launcher
-  MG42

**Comment by Liruxo** [comment]
@rolylolypoly Well as you said i dont have any problems with this gradle build on linux, so i cant help you with windows problems. Anyways this sounds very suspicious since gradle, git and java should work the same on any operating system

**Review by Liruxo — COMMENTED** [review]
To anyone reading this, just ignore it i was testing something and i cant delete this

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java
Comment (jamioflan) [inline src/main/java/com/flansmod/common/driveables/EntityDriveable.java]: Can't we just catch it, error and return?

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java
Comment (jamioflan) [inline src/main/java/com/flansmod/common/driveables/EntityDriveable.java:787]: Could this name be more descriptive? ShotFiredSuccessHandler or something? It doesn't say which bit of shooting is being handled

File: src/main/java/com/flansmod/common/driveables/mechas/EntityMecha.java
Comment (jamioflan) [inline src/main/java/com/flansmod/common/driveables/mechas/EntityMecha.java:395]: Mechas definitely can shoot grenades. I'm happy to put this PR in with some TODOs in though as we need to get some progress on it

File: src/main/java/com/flansmod/common/guns/GunType.java
Comment (jamioflan) [inline src/main/java/com/flansmod/common/guns/GunType.java:250]: Has this been tested on dedicated servers? Turning on some of this stuff can cause runtime classNotFoundExceptions

File: src/main/java/com/flansmod/common/guns/ShotHandler.java
Comment (jamioflan) [reply src/main/java/com/flansmod/common/guns/ShotHandler.java:162]: My bad. I started writing C++ and C# between starting Flan's Mod and some of the later updates, so my style is a bit mixed

**Review by jamioflan — APPROVED** [review]
Happy to take this as is. The number of outstanding TODOs is no worse than what's already in the source, and a review this big is likely to take forever unless we get it in.

**Comment by jamioflan** [comment]
Also @Liruxo, can you send an e-mail to admin@flansmod.com? I'd like to send you a Discord invite

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/driveables/EntityDriveable.java]: I have inserted a casting check for good coding practices, but i plan on reworking grenades anyways which would completely resolve all problems related to them

File: src/main/java/com/flansmod/common/driveables/EntityDriveable.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/driveables/EntityDriveable.java:787]: Well it is just a handler called every time when a bullet is fired, so ShotFiredSuccessHandler doesn't really make sense to me. I have added a description in the interface, this may help understanding the function of the interface

File: src/main/java/com/flansmod/common/driveables/mechas/EntityMecha.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/driveables/mechas/EntityMecha.java:395]: As far as my testing goes they can not use grenade launchers (#1023). I am currently unable to test this due to #1060, but i don't think anything changed

File: src/main/java/com/flansmod/common/guns/GunType.java
Comment (Liruxo) [reply src/main/java/com/flansmod/common/guns/GunType.java:250]: since i removed it, i don't think that there should be any problem concerning this. But i tested all of this on a server so there is definitely no problem

**Pull request merged**
